### PR TITLE
Loss v len callback

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
Adds the loss v len callback that logs loss vs context length for each iteration into tables. Using those tables we can plot graphs such as the following:

Screenshot 2024-05-22 at 2 58 54 PM

### CODE REVIEW

File: llmfoundry/callbacks/loss_perp_v_len_callback.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+    def __init__(
+        self,
+        ignore_index: int,
+        dist_sync_on_step: bool = False,
+    ):
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+
+        self.needs_batch = True
+        self.ignore_index = ignore_index
+        self.loss_fn = torch.nn.CrossEntropyLoss(
```
Comment (vchiley) [inline llmfoundry/callbacks/loss_perp_v_len_callback.py:168]: Is it possible to use the model loss fn instead of torch.nn.CrossEntropyLoss
does this make senes?

Comment (ShashankMosaicML) [reply llmfoundry/callbacks/loss_perp_v_len_callback.py:168]: Done.

File: llmfoundry/callbacks/loss_perp_v_len_callback.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    Args:
+        log_batch_interval (int): The interval for logging.
+        compute_batch_interval (int): The interval for computing the metric.
+        ignore_index (int): Specifies a target value that is ignored for computing loss.
+    """
+
+    def __init__(
+        self,
+        log_batch_interval: int,
+        compute_batch_interval: int,
```
Comment (vchiley) [inline llmfoundry/callbacks/loss_perp_v_len_callback.py:31]: Should we add a comment here saying that eventually we should consider making this into one arg?
(once that other issue gets resolved)

File: llmfoundry/callbacks/loss_perp_v_len_callback.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+from llmfoundry.models.mpt import ComposerMPTCausalLM
+
+__all__ = [
+    'LossPerpVsContextLengthLogger',
+]
+
+
+class LossPerpVsContextLengthLogger(Callback):
+    """Logs the average loss and perplexity for every context length.
```
Comment (vchiley) [inline llmfoundry/callbacks/loss_perp_v_len_callback.py:20]: Should we note that this only works with MLflow? (Is that true or am I misremembering something?)

File: llmfoundry/callbacks/loss_perp_v_len_callback.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+from torchmetrics import Metric
+
+from llmfoundry.models.mpt import ComposerMPTCausalLM
+
+__all__ = [
+    'LossPerpVsContextLengthLogger',
+]
+
+
+class LossPerpVsContextLengthLogger(Callback):
```
Comment (dakinggg) [inline llmfoundry/callbacks/loss_perp_v_len_callback.py:19]: @experimental_class

File: llmfoundry/callbacks/loss_perp_v_len_callback.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+                'log_batch_interval is shorter than the compute_batch_interval for LossPerpVsContextLengthLogger.',
+            )
+        self.log_batch_interval = log_batch_interval
+        self.compute_batch_interval = compute_batch_interval
+        self.ignore_index = ignore_index
+        self.metric_dict = {}
+        self.loss_perp_v_len = LossPerpVLen(ignore_index)
+
+    def after_backward(self, state: State, logger: Logger) -> None:
+        if not isinstance(state.model, ComposerMPTCausalLM):
```
Comment (dakinggg) [inline llmfoundry/callbacks/loss_perp_v_len_callback.py:45]: this check should happen on the `init` event so that it fails fast

File: llmfoundry/callbacks/loss_perp_v_len_callback.py (hunk 4 above)
Comment (ShashankMosaicML) [reply llmfoundry/callbacks/loss_perp_v_len_callback.py:21]: Done.

File: llmfoundry/callbacks/loss_perp_v_len_callback.py (hunk 3 above)
Comment (ShashankMosaicML) [reply llmfoundry/callbacks/loss_perp_v_len_callback.py:22]: Done.

File: llmfoundry/callbacks/loss_perp_v_len_callback.py (hunk 2 above)
Comment (ShashankMosaicML) [reply llmfoundry/callbacks/loss_perp_v_len_callback.py:33]: Done.

File: llmfoundry/callbacks/loss_perp_v_len_callback.py (hunk 5 above)
Comment (ShashankMosaicML) [reply llmfoundry/callbacks/loss_perp_v_len_callback.py:49]: Done.

**Review by dakinggg — APPROVED** [review]
